Packaging improvements

0pcom opened the pull request:
Did you run `make format && make check`?
no
Fixes #	
* autoconfiguration of visor and hypervisor upon package installation

 Changes:	
* generates ssl key and cert and adds them to the package at the location they are found at in skybian following the procedure detailed [here](https://github.com/SkycoinProject/skywire-mainnet/wiki/Skywire-Mainnet-Installation-From-Source#enable-tls)
* adds `postinst` script which is automatically executed upon package instalation to accomplish configuration of hypervisor and visor.
* adds package-based method of setting / distributing the hypervisor public key
* adds local apt repository configuration for distributing the hypervisorconfig package
* hosts the apt repo on the local network with `readonlycache`

 Notes:
* the hypervisor configuration does not take place if the `hypervisorconfig` package is installed
* the `hypervsorconfig` package is created in the `postinst` script when the hypervisor is configured, but it is not installed on the hypervisor. It is added to the hypervisor's apt repo in /var/cache/apt/repo
* the configuration of the apt repo hosted by the hypervisor in the visor's /etc/apt/sources.list **is not included in these scripts.**

This should be done either manually by the user in a ssh session or via a functionality which can be added to skyimager. Example:
```
add-apt-repository 'deb  [trusted=yes] http://192.168.0.2 sid main'
```
Substituiting in the above command the debian codename which is listed in `/etc/os-release`

Issues:
* the visor fails to start when the hypervisor key is set in /etc/skywire-visor.json
```
failed to parse config. error="json: cannot unmarshal object in go struct field V1.hypervisors of type cypher.Pubkey"
```
This does not happen with the master branch, only with develop.

Conflicts:
The extent of the configuration which is accomplished by the postinst script __may cause difficulty__ when it is installed in a chroot in skybian. 

How to test this PR:
```
make package-amd64
dpkg -i *.deb
```
go to 127.0.0.1:8000

hypervisor is running, visor __would be running__ if not for the aforementioned errors.

nkryuchkov commented on cmd/skywire-node/commands/root.go:
```diff
@@ -16,97 +18,173 @@ import (
 	"github.com/skycoin/skycoin/src/util/logging"
 	"github.com/spf13/cobra"
 
+	"net/http"
```
Could you please move this import and the below one up to the first import block in order to have them all in one block?

ayuryshev replied:
Indeed. I forgot `make format` before push.
But not for this file. I never format by hand import in golang sources - it's too fragile.

nkryuchkov commented on cmd/skywire-node/commands/root.go:
```diff
@@ -16,97 +18,173 @@ import (
+		option = profile.BlockProfile
+	case "trace":
+		option = profile.TraceProfile
+	default:
```
Could you consider removing empty `default` case here?

ayuryshev replied:
Yes

nkryuchkov commented on cmd/skywire-node/commands/root.go:
```diff
@@ -16,97 +18,173 @@ import (
+	if cfg.syslogAddr != "none" {
+		hook, err := logrus_syslog.NewSyslogHook("udp", cfg.syslogAddr, syslog.LOG_INFO, cfg.tag)
+		if err != nil {
+			cfg.logger.Error("Unable to connect to syslog daemon")
```
I'd also log the returned error here

ayuryshev replied:
Agree

nkryuchkov commented on cmd/skywire-node/commands/root.go:
```diff
@@ -16,97 +18,173 @@ import (
+func (cfg *runCfg) startProfiler() *runCfg {
+	var option func(*profile.Profile)
+	switch cfg.profileMode {
+	case "none":
```
What do you think about creating a `const` block and probably a new type with all possible values for `profileMode` in order to avoid comparison with string literals?

ayuryshev replied:
If I was a developer of pkg/profile - I would concider this way.
But that's part of theirs API and recommended way to use this package.
Look at https://github.com/pkg/profile/blob/f6fe06335df110bcf1ed6d4e852b760bfc15beee/example_test.go#L42

IMO: API provide by package is "good enough". I don't see the point to improve it inside skywire-node.

nkryuchkov commented on cmd/skywire-node/commands/root.go:
```diff
@@ -16,97 +18,173 @@ import (
+	return cfg
+}
+
+// waitOsSignalsAndDelay
```
Is there any purpose for this comment?

ayuryshev replied:
removed

nkryuchkov commented on cmd/skywire-node/commands/root.go:
```diff
@@ -16,97 +18,173 @@ import (
+func (cfg *runCfg) waitOsSignals() *runCfg {
+	ch := make(chan os.Signal, 2)
+	signal.Notify(ch, []os.Signal{syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT}...)
+	<-ch
```
Could you please explain in short why is reading from `ch` used here?

ayuryshev replied:
It's a common patern in using os.signal:
- register signals 
- block execution waiting for signal
Look at examples https://golang.org/pkg/os/signal/#Notify

As for `make(chan os.Signal, 2)` - buffer size=2 is inherited. 
Maybe `1` will be sufficient. Not tested

nkryuchkov replied:
@ayuryshev Yeah, it's a common pattern. I just don't fully understand why it happens before starting a goroutine, which tries to read from the channel again. Could this line be removed here and the code in the `select` below be unwrapped from the goroutine in order that the timeout works? 

However, I might incorrectly understand this code.

ayuryshev replied:
It starts goroutine that allow skywire-node to shutdown in `cfg.conf.ShutdownTimeout` period of time or make it possible to kill skywire-node with second signal.

So it makes possible: "Ctrl-C [wait for shutdown]" or "Ctrl-C Ctrl-C"

nkryuchkov replied:
@ayuryshev thanks for the explanation. I misunderstood the purpose of the method.

nkryuchkov commented on internal/httpauth/client_test.go:
```diff
@@ -6,14 +6,17 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"github.com/skycoin/skycoin/src/util/logging"
```
What do you think of swapping this line and the above line because `skycoin/skycoin` is another repository?

ayuryshev replied:
I pushed code after `make format`. 
It could be outdated or not - IMO such things must be solved automatically

jdknives committed:
Merge pull request #349 from SkycoinProject/fix/hypervisor-reconnect-#346

Fix hypervisor disconnection issue.

Darkren committed:
Add app API docs

Darkren committed:
Experiment with table of contents

Darkren committed:
Refactor docs a bit

Darkren committed:
Refactor a bit more

jdknives committed:
Merge pull request #353 from Darkren/feature/app-api-docs

Add app API docs

Darkren committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into feature/vpn

Darkren committed:
Replace local `netutil` with the one from `dmsg`

Darkren committed:
Utilize skyenv in apps

Darkren committed:
Simplify IP to octets routine

Darkren committed:
Remove redundant logs

Darkren committed:
Try to remove binary route from ifconfig call

Darkren committed:
Add debug logs

Darkren committed:
Fix PATH once again

Darkren committed:
Pass PATH env for VPN server too

Darkren committed:
Remove redundant logs

Darkren committed:
Add all cmds called by path relative to PATH

Darkren committed:
Merge pull request #278 from Darkren/feature/vpn

VPN apps

Darkren committed:
Add passcode to config generation

Darkren committed:
FUlly add passcode

Darkren committed:
Fix `String` for `HandshakeStatus`

Darkren committed:
Add debug log

Darkren committed:
Fix passcode passing to the server

Darkren committed:
Remove debug log

Darkren committed:
Make lintr happy

Darkren committed:
Remove redundant logs

Darkren committed:
Add RPC calls to change passcode and PK for VPN apps

Darkren committed:
Make linter happy

Darkren committed:
Add `README.md` for VPN client

Darkren committed:
Add `README.md` for VPN server, fix client's one a bit

Darkren committed:
Make default network gateway detection work via `ip r` for Linux systems

Darkren committed:
Fix default network gateway detection

Darkren committed:
Fix it once again

Darkren committed:
Make TUN setup go via `ip` for Linux systems

Darkren committed:
Fix Linux code

Darkren committed:
Try to fix

Darkren committed:
Make all routing be done via 'ip` for Linux systems

Darkren committed:
Remove some duplicate code

Darkren committed:
Substitute logrus sys log hook with our wrapper

Darkren committed:
Implement `DefaultNetworkGateway` for Windows

Darkren committed:
Finish TUN setup for Windows

Darkren committed:
Fix sys log build for Windows

Darkren committed:
Merge branch 'feature/vpn-improvements' of https://github.com/Darkren/skywire-mainnet into feature/vpn-improvements

Darkren committed:
Refactor all os-related funcs

Darkren committed:
Fix `DefaultNetworkInterface` for Linux systems

Darkren committed:
Add routing commands for Windows

Darkren committed:
Merge branch 'feature/vpn-improvements' of https://github.com/Darkren/skywire-mainnet into feature/vpn-improvements

Darkren committed:
Remove dmsgpty config generation for Windows subsystems

Darkren committed:
Remove dmsgpty references from visor package

Darkren committed:
Let it be built for Windows systems

Darkren committed:
Fix dmsgpty setup for Windows

Darkren committed:
Fix it once again

Darkren committed:
Try to use TAP server side for Windows clients

Darkren committed:
Fix

Darkren committed:
Fix

Darkren committed:
Fix

Darkren committed:
Fix

Darkren committed:
Fix

Darkren committed:
Start integrating wintun

Darkren committed:
Udate vendor

Darkren committed:
Revert testing changes

Darkren committed:
Finish wintun integration

Darkren committed:
Add a bit TODOs

Darkren committed:
Refactor VPN part

Darkren committed:
Refactor main CMDs and syslog

Darkren committed:
Refactor hypervisor

Darkren committed:
Refactor visor

Darkren committed:
Make linter happy

Darkren committed:
Update vendor

Darkren committed:
Try to integrate noise

Darkren committed:
Add more logging

arc1999 committed:
retain keys

arc1999 committed:
formating done

arc1999 committed:
formating done

jdknives committed:
Merge pull request #359 from arc1999/feature/retain-keys

Feature/retain keys

Darkren committed:
Improve logging

Darkren committed:
Fix

Darkren committed:
Fix

Darkren committed:
Fix

Darkren committed:
Pass PK and SK as binary args

Darkren committed:
Add noise handshake

Darkren committed:
Refactor encryption

Commit:
Appserver and visor improvements. (#357)

Appserver improvements:
* Combined the appserver.Server and appserver.ProcManager concepts.
* Have one rpc.Server per app proc.
* Simplified app.Client initialization.
* Visor to App initialization is done via a single env defined for the app PROC_CONFIG which contains a JSON object.
* App to Visor initialization is done by App connecting the Visor via app server address. The first 16 bytes that the App sends to Visor is the appcommon.ProcKey.
* Changed appcommon.ProcKey format so that it is a deterministic length.
* Improved app logging and removed any possibilities of log database race conditions with bbolt.
* Various logging improvements.

App management, visor config, startup and shutdown improvements:
* Separated app-associated functionality within /pkg/visor into /pkg/app/launcher.
* Added the concept of visor.BaseConfig() to simplify obtaining default values.
* Made startup/shutdown logic modular and introduced init and shutdown stacks.
* Shutdown now uses module-based timeouts and wait groups.

Changes as suggested by @Darkren and @nkryuchkov

evanlinjin committed:
Initial work on versioned visor configs.
* Moved visor conifg-related parts to a new module /pkg/visor/visorconfig.
* Introduced visorconfig.Common that contains shared fields across all config versions.
* Improved config-parsing logic.

evanlinjin committed:
Fixed tests.

evanlinjin committed:
Clean up /cmd/skywire-visor/commands/root.go

evanlinjin committed:
Fixed visor config generation.
* Simplifications to 'skywire-cli visor gen-config'.

Darkren committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into feature/vpn-improvements

Darkren committed:
Fix VPN envs

evanlinjin committed:
Invert shutdown order.
* We should shut down the module with the the most dependencies first.

evanlinjin committed:
Re-implemented testenv config generation (#365)

evanlinjin committed:
Implement config upgrading from v0.0.0

evanlinjin committed:
Added visorconfig tests.

Darkren committed:
Remove panic

Commit:
add package directives to Makefile; add systemd services for visor and hypervisor

evanlinjin committed:
Added public key (pk) back into visor config.

evanlinjin committed:
Changes as suggested by @nkryuchkov

Darkren committed:
Make linter happier

evanlinjin committed:
Fixed logging level settings.

Darkren committed:
Add Windows info to README

evanlinjin committed:
Format as suggested by @Darkren

evanlinjin committed:
Ran: go generate.

Darkren committed:
Exclude dmsgpty config generation for Windows systems

Darkren committed:
Disallow socket files removal for Windows systems

jdknives committed:
Merge pull request #364 from SkycoinProject/feature/versioned-config-#362

Versioned visor configs.

Darkren committed:
Add an ability to derive PK out of SK for noise creds

Darkren committed:
Refactor a bit

Darkren committed:
Refactor, make VPN server autostart velue false by default

Darkren committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into feature/vpn-improvements

evanlinjin committed:
Appevent module implementation.

The appevent module allows for events to be broadcasted from visor (via appserver) to apps.
Events have an event type what determines the contents of the event data.
Apps subscribe to chosen event types, which are broadcasted via event channels (one event channel per event type).

* Introduced a new optional line of communication between app/visor for event broadcast, in which app hosts the listener and the visor (via appserver) does the dial. The app advertises this optional listening address (if it exists) to the visor via the main app <-> visor connection.
* Apps now use the appcommon.Hello JSON object as their initial dial message to appserver (this contains the proc_key, and event broadcast connection related fields).
* app.NewClient() now has an optional appevent.Subscriber as input. If defined, the broadcast event connection will be created and the appevent.Subscriber object will be used for receiving events.
* Updated dmsg version to make use of callbacks (for dialing and closing sessions).

Other changes:
* Updated golangci-lint version for travis.
* Updated helloworld app to listen/write in a loop and to log all events comming from appserver.

Commit:
Merge pull request #356 from Darkren/feature/vpn-improvements

VPN improvements

Darkren committed:
Make everything generic on client side

Darkren committed:
Rename client package itself

Darkren committed:
Includ type to the sercice entry

Darkren committed:
Finish?

evanlinjin committed:
Changed skywire-visor help to be consistent with actual behavior.

Darkren committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into feature/make-proxy-disc-generic

Darkren committed:
Make linter happy, add VPN disc client to the factory

evanlinjin committed:
Changes as suggested by @Darkren
* Remove various empty lines and fix typos.
* Remove unnecessary chan close in appevent.Broadcaster.
* Make multiple calls to .OnTCPDial and .OnTCPClose safe for appevent.Subscriber

Darkren committed:
Merge pull request #371 from SkycoinProject/feature/app-rpc-interface

Appevent module implementation.

evanlinjin committed:
Wait for dmsg network during visor startup.

evanlinjin committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into develop

evanlinjin committed:
Fix transport manager hangs and various improvements (#375).
* Restructured (transport.Manager).SaveTransport to improve readability and ensure we always close the managed transport before calling mTp.wg.Wait(). In fact, we are now using mTp.Close() directly (which does both).
* Visor's RPC client timeout is now optional. Setting a value of 0 duration will result in no timeout.
* Disabled visor RPC client timeout for visor CLI.

evanlinjin committed:
Changes as suggested by @Darkren and @nkryuchkov

* Replaced goto statement with for loop in transport.Manager

evanlinjin committed:
Merge pull request #376 from SkycoinProject/fix/tpm-hangs-#375

Fix transport manager hangs and various improvements.

nkryuchkov committed:
Move buildinfo to dmsg repo

nkryuchkov committed:
Fix dmsg version

jdknives committed:
Merge pull request #377 from nkryuchkov/fix/move-buildinfo

Move buildinfo to dmsg repo

nkryuchkov committed:
Implement a holepunch transport (#313)

* Implement `stcp` with address resolving (`stcpr`) and encryption
* Implement `stcp` with TCP hole punching (`stcph`) and encryption
* Add encryption for all `stcp`

nkryuchkov committed:
Updater improvements

nkryuchkov committed:
Accept visor config from args without setting --config

nkryuchkov committed:
Add visor delay flag back

nkryuchkov committed:
Disable releases for some platforms temporarily

nkryuchkov committed:
Add delay flag to hypervisor

nkryuchkov committed:
Fix missing build info in releases

nkryuchkov committed:
Update VPN apps

nkryuchkov committed:
Specify stable/testing channels for updating

nkryuchkov committed:
Add VPN apps to updater

nkryuchkov committed:
Fix updating if release contains new apps

nkryuchkov committed:
Fix release targets

evanlinjin committed:
Improve reliability and logging of setup node.
* Split setup node into multiple functions for easier testability.
* Check inputs for each setup request.
* Use context.Context in all parts of the setup process to ensure we cancel on hard timeout.
* Introduced routerclient.Map to clean up various parts of the logic, as well as allow the setup process to be speedier.
* Got rid of the routing.Path data type as it is confusing. I has resorted to use []routing.Hop directly.
* Unit tests.

0pcom committed:
Update Makefile

Darkren committed:
Resolve queries

jdknives committed:
Merge pull request #374 from Darkren/feature/make-proxy-disc-generic

Make proxy disc generic

evanlinjin committed:
Merge from 'origin/develop'.

nkryuchkov committed:
Remove --delay flag

nkryuchkov committed:
Add lint-extra target to Makefile

nkryuchkov committed:
Fix some linter complaints

nkryuchkov committed:
Extract updating hypervisor to a separate endpoint

nkryuchkov committed:
Format imports

nkryuchkov committed:
Fix updater behavior for systemd

nkryuchkov committed:
Skip binary if it is not found is update

jdknives committed:
Merge pull request #384 from nkryuchkov/feature/format-imports

Set up formatting imports using goimports-reviser

jdknives committed:
Fix/use vendor (#329)

* add -mod=vendor flag to tests

* add dmsgpty vendor

* remove accidental change

* Vendor

Co-authored-by: Nikita Kryuchkov <nkryuchkov10@gmail.com>

nkryuchkov committed:
Merge branch 'develop' into feature/updater-improvements

nkryuchkov committed:
Add dummy delay flag

nkryuchkov committed:
Merge remote-tracking branch 'nkryuchkov/feature/updater-improvements' into feature/updater-improvements

nkryuchkov committed:
Fix passing config to visor

nkryuchkov committed:
Workaround for config

nkryuchkov committed:
Fix config converter

nkryuchkov committed:
Disable STCPR and STCPH if address-resolver is unreachable

evanlinjin committed:
Changes as suggested by @nkryuchkov

* Rewrote setup.Node tests.

* Various tweaks/renamings.

evanlinjin committed:
Merge branch 'develop' into fix/setup-node-hangs

evanlinjin committed:
Format code.

Commit:
revisions for PR#366

0pcom committed:
Merge branch 'develop' into develop

nkryuchkov committed:
Add --delay flag to visor and hypervisor for backwards compatibility

nkryuchkov committed:
Try to update from systemd and old version correctly

nkryuchkov committed:
Try to update from systemd and old version correctly

Commit:
move dPKGBUILD.sh ro scripts and change makefile accordingly.

Commit:
move dPKGBUILD.sh ro scripts and change makefile accordingly.

0pcom committed:
Delete dPKGBUILD.sh

nkryuchkov committed:
Try to update from systemd and old version correctly

nkryuchkov committed:
Fix updater bug

nkryuchkov committed:
Try to update from systemd and old version correctly

nkryuchkov committed:
Try to update from systemd and old version correctly

nkryuchkov committed:
Add comments for clarity

nkryuchkov committed:
Fix comments

nkryuchkov committed:
Fix comments

Commit:
cleanup

nkryuchkov committed:
Fix algorithm of checking if parent process is run by systemd

nkryuchkov committed:
Fix vendor

nkryuchkov committed:
Fix logging

Darkren committed:
Fix updater interval panic

0pcom committed:
fix small typo

nkryuchkov committed:
Rename proxy discovery to service discovery

evanlinjin committed:
Tweaks as suggested by @nkryuchkov

Actual dialing to the snettest.Env struct has also been temporarily removed. This will be addressed in #395

Darkren committed:
Add TODO

Darkren committed:
Rename `proxyUpdater` => `serviceUpdater`

Darkren committed:
Make address resolvers non-required to run VPN client

nkryuchkov committed:
Merge pull request #381 from SkycoinProject/fix/setup-node-hangs

Improve reliability and logging of setup node.

nkryuchkov committed:
Change domains

Darkren committed:
Add debug logs

Darkren committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into fix/vpn-client

Darkren committed:
Fix debug logs

Darkren committed:
Fix

nkryuchkov committed:
Implement a UDP transport

Darkren committed:
Add binaries

nkryuchkov committed:
Merge branch 'develop' into feature/sudp

# Conflicts:
#	go.mod
#	go.sum
#	pkg/snet/network.go
#	pkg/snet/snettest/env.go
#	vendor/modules.txt

Darkren committed:
Fix VPN client launch from hypervisor

nkryuchkov committed:
Remove redundant log message

Commit:
update makedepends to include python and python2

Commit:
fix missing variable

Darkren committed:
Revert .gitignore changes

Darkren committed:
Sort out some PR queries

4rchim3d3s committed:
Create de.json

4rchim3d3s committed:
Add de_base.json

4rchim3d3s committed:
Add de.png

4rchim3d3s committed:
add german language

4rchim3d3s committed:
update de.json

jdknives committed:
Merge pull request #390 from Darkren/fix/vpn-client

Fix VPN apps

nkryuchkov committed:
Add SUDP to config generation

jdknives committed:
Merge pull request #393 from nkryuchkov/fix/rename-proxy-discovery

Rename proxy discovery to service discovery

nkryuchkov committed:
Fix suggested by @Darkren

nkryuchkov committed:
Fix sync.WaitGroup bug

nkryuchkov committed:
Merge pull request #397 from nkryuchkov/feature/sudp

Implement a UDP transport

nkryuchkov committed:
Merge pull request #380 from nkryuchkov/feature/updater-improvements

Updating improvements

Darkren committed:
Make visor retyr dialing TP disc till success

Darkren committed:
Refactor imports

Darkren committed:
Adjust logging

jdknives committed:
rebuild frontend

Darkren committed:
Ensure that all attempts to dial or listen to stcpr/stcph are protected in terms of not-yet-ready client

Darkren committed:
Almost finish mutexes dance

Darkren committed:
Make mutexes RW

Darkren committed:
Small fix

Senyoret1 committed:
Fix a problem with the app names in the manager

jdknives committed:
Merge pull request #403 from Senyoret1/app-name

Fix a problem with the app names in the manager

jdknives committed:
merge develop

Darkren committed:
Try to fix transport initialization

Darkren committed:
Add debug logs

Darkren committed:
Add more debug logs

Darkren committed:
Fix networks initialization

Darkren committed:
Remove debug logs

Darkren committed:
Improve logging

Darkren committed:
Improve logging

Commit:
Fixes for #366; also add sudo to makedepends; add create binary checksums.txt during build.

0pcom committed:
Merge branch 'develop' into develop

Darkren committed:
Finish, cleanup

Darkren committed:
Merge branch 'develop' of github.com:SkycoinProject/skywire-mainnet into feature/allow-visor-to-start-without-tp-disc-and-addr-resolver

Darkren committed:
Update vendor

4rchim3d3s committed:
change wording and fault correction

4rchim3d3s committed:
change word "Apps" to "Anwendungen"

Darkren committed:
Refactor, fix queries

jdknives committed:
Merge pull request #366 from Skyfleet/develop

Add package directive to Makefile

jdknives committed:
fix conflicts

jdknives committed:
conflicts fixed

jdknives committed:
Merge pull request #399 from 4rchim3d3s/patch-1

Added German Language

Senyoret1 committed:
Add a copy button in the visor list

jdknives committed:
Merge pull request #406 from Senyoret1/ui-changes

Add a copy button in the visor list

jdknives committed:
Merge pull request #402 from Darkren/feature/allow-visor-to-start-without-tp-disc-and-addr-resolver

Allow visor to start in case of address resolver or tp disc failure
